utils/zerovec/src/ule/niche.rs

```diff
@@ -0,0 +1,184 @@
+    }
+}
+
+impl<U: NicheBytes<N> + ULE, const N: usize> From<NichedOptionULE<U, N>> for Option<U> {
+    fn from(noule: NichedOptionULE<U, N>) -> Self {
+        NichedOptionULE::get(noule)
+    }
+}
+
+impl<U: NicheBytes<N> + ULE, const N: usize> Copy for NichedOptionULE<U, N> {}
+
+impl<U: NicheBytes<N> + ULE, const N: usize> Clone for NichedOptionULE<U, N> {
+    fn clone(&self) -> Self {
+        *self
+    }
+}
+
+impl<U: NicheBytes<N> + ULE + PartialEq, const N: usize> PartialEq for NichedOptionULE<U, N> {
+    fn eq(&self, other: &Self) -> bool {
+        self.get().eq(&other.get())
+    }
+}
+
+impl<U: NicheBytes<N> + ULE + Eq, const N: usize> Eq for NichedOptionULE<U, N> {}
+
+/// Safety for ULE trait
+/// 1. NichedOptionULE does not have any padding bytes due to `#[repr(packed)]` on a struct
+///    containing only ULE fields.
+///    NichedOptionULE either contains INVALID_BIT_PATTERN or valid U byte sequences.
+///    In both cases the data is initialized.
+/// 2. NichedOptionULE is aligned to 1 byte due to `#[repr(packed)]` on a struct containing only
+///    ULE fields.
+/// 3. validate_byte_slice impl returns an error if invalid bytes are encountered.
+/// 4. validate_byte_slice impl returns an error there are extra bytes.
+/// 5. The other ULE methods are left to their default impl.
+/// 6. NichedOptionULE equality is based on ULE equality of the subfield.
+unsafe impl<U: NicheBytes<N> + ULE, const N: usize> ULE for NichedOptionULE<U, N> {
+    fn validate_byte_slice(bytes: &[u8]) -> Result<(), crate::ZeroVecError> {
+        let size = size_of::<Self>();
+        // The bytes should fully transmute to a collection of Self
+        if bytes.len() % size != 0 {
+            return Err(crate::ZeroVecError::length::<Self>(bytes.len()));
+        }
+        bytes.chunks(size).try_for_each(|chunk| {
+            // Associated const cannot be referenced in a pattern
+            // https://doc.rust-lang.org/error-index.html#E0158
+            if chunk == <U as NicheBytes<N>>::INVALID_BIT_PATTERN {
+                Ok(())
+            } else {
+                U::validate_byte_slice(chunk)
+            }
+        })
+    }
+}
+
+/// Optional type which uses [`NichedOptionULE<U,N>`] as ULE type.
+/// Derefs to [`Option<U>`]
+/// The implementors guarantee that N == core::mem::sizeo_of::<Self>()
+/// [`repr(transparent)`] guarantees that the layout is same as [`Option<U>`]
+#[derive(Debug, Copy, Clone, PartialEq, Eq, Hash)]
+#[repr(transparent)]
+pub struct NichedOption<U, const N: usize>(Option<U>);
+
+impl<U, const N: usize> NichedOption<U, N> {
+    pub const fn new(o: Option<U>) -> Self {
+        Self(o)
+    }
+}
+
+impl<U, const N: usize> Default for NichedOption<U, N> {
+    fn default() -> Self {
+        Self(None)
+    }
+}
+
+impl<U, const N: usize> From<Option<U>> for NichedOption<U, N> {
+    fn from(o: Option<U>) -> Self {
+        Self(o)
+    }
+}
+
+impl<U, const N: usize> Deref for NichedOption<U, N> {
```

Comment:
issue: this is an incorrect usage of Deref

Reply:
Removing `Deref` and `DerefMut` implementations for now.

Reply:
Why is this an incorrect usage of Deref?

Reply:
Deref is supposed to be used in cases where there's actual pointery indirection happening, not just to get the benefit of deref coercions.

Reply:
from the deref docs:

> Implementing Deref for smart pointers makes accessing the data behind them convenient, which is why they implement Deref. On the other hand, the rules regarding Deref and [DerefMut](https://doc.rust-lang.org/1.23.0/std/ops/trait.DerefMut.html) were designed specifically to accommodate smart pointers. Because of this, Deref should only be implemented for smart pointers to avoid confusion.

Reply:
ok. I do think this is a convention that should be revisited on a broader level. But I'll buy that this isn't the place to shake it up.